Closes ulsdevteam/pkp-betterPassword#15: Added password expiration, disallowed reuse and small fixes.

jonasraoni opened the pull request:
Hi @ctgraham,

Please take a look if the changes are ok, I've checked against an OJS 3.2.1.

The password expiration is "friendly", the user won't be disturbed until the next time the tries to login.

I'll try to leave some comments here and there to help.

jonasraoni reviewed (commented):
Just left some comments :)

jonasraoni commented on BetterPasswordSettingsForm.inc.php:
```diff
@@ -37,16 +37,26 @@ function __construct($plugin, $contextId) {
 			}
 		}
-
+ __('plugins.generic.betterPassword.manager.settings.betterPasswordLockRequired');
```
Forgotten line (I was checking the locale).
I've already updated locally the typos that I saw, but I'll wait your comments before pushing updates.

ctgraham reviewed (changes requested):
Thanks, @jonasraoni!  This is a great addition, with a lot of drive by cleanup as well.  I've added a few requests.  Happy 2021 to you!

jonasraoni commented on BetterPasswordPlugin.inc.php:
```diff
@@ -35,8 +35,9 @@ class betterPasswordPlugin extends GenericPlugin {
 		'minPasswordLength' => 'int',
-	);
-	
+		'betterPasswordInvalidationSeconds' => 'int',
```
I was thinking about it too... I'd like to be neutral and use an interval component (the setting could be saved using a standard format, e.g. `P2H30M`).

So well, I'll update to use days!

jonasraoni commented on BetterPasswordPlugin.inc.php:
```diff
@@ -327,24 +366,46 @@ function handleTemplateDisplay($hookName, $args) {
+					return true;
+				}
+			}
+			if ($passwordLifetime && $user->getLastPasswordUpdate() + $passwordLifetime < time()) {
```
I agree! I'll check the size of my todo list, maybe I can add it to the list :)

jonasraoni commented on BetterPasswordPlugin.inc.php:
```diff
@@ -327,24 +366,46 @@ function handleTemplateDisplay($hookName, $args) {
+					return true;
+				}
+			}
+			if ($passwordLifetime && $user->getLastPasswordUpdate() + $passwordLifetime < time()) {
```
We were going to send custom emails, but the feature was cancelled... I'll add this!

jonasraoni commented on xml/schema.xml:
```diff
@@ -22,7 +22,8 @@ betterPassword plugin schema in XML
         <field name="failed_login_time" type="T" >
             <NOTNULL/>
         </field>
+        <field name="passwords" type="X" />
```
Yeah, I was going to use the settings, but as the plugin already had its own table, I thought it would be better to drop my data there as well, to decrease the pollution/orphan data.

ps: I'm curious if there's something to uninstall the plugins.

jonasraoni replied:
I've replaced the fields by the user setting.

jonasraoni reviewed (commented):
Just left some new questions, I'll start fixing things probably tomorrow :)

ctgraham commented:
@jonasraoni there is a small bugfix (b940efd) against `master` for 3.2 to work with password resets.  You may want to pull from this repo to bring it into your branch.

jonasraoni commented:
@ctgraham Thanks for the warning! I'm trying to clear my task stack to get back to this PR :D

jonasraoni commented on BetterPasswordPlugin.inc.php:
```diff
@@ -283,16 +323,15 @@ function getBlacklists() {
 		$userBlacklists = $this->getSetting(CONTEXT_SITE, 'betterPasswordUserBlacklistFiles');
-		$userBlacklistsFilenames = array_keys($userBlacklists);
-		foreach ($userBlacklistsFilenames as $f) {
+		foreach ($userBlacklists as $f) {
```
Yeah, I had issues with the hashes. Some places were using the hash of the filename and others the hash of its content.

I'll just grab the `is_array($userBlacklists)` fix from this PR.

jonasraoni commented on BetterPasswordPlugin.inc.php:
```diff
@@ -327,24 +366,46 @@ function handleTemplateDisplay($hookName, $args) {
+		$maxReusablePasswords = $this->getSetting(CONTEXT_SITE, 'betterPasswordInvalidationPasswords');
+		$passwordLifetime = $this->getSetting(CONTEXT_SITE, 'betterPasswordInvalidationSeconds');
+
+		if ($args[0] === "tab.user.ProfileTabHandler" && $args[1] === 'savePassword' && $maxReusablePasswords) {
```
Here there was an issue! There are two interfaces to update the password and I'm covering just one...
I've fixed.

jonasraoni commented:
Hi @ctgraham! I think I've addressed all the issues, also I've tried to decrease the indentation by returning early.

I've tested the features one by one, but it would be better if someone could double-check 👀

I've changed the definition of blacklist to blocklist, but I've left some things with `blacklist`... If you're ok to proceed with the break (and possibly prepare a migration script), then I would also rename the table from `badpw_failedlogins` to maybe `better_password_failed_attempts` 🤔

One source conflict is that I've changed how the blocklist is saved, now it's an array in the form: `[hash => filename]`, so the user can files with the same name, but different content.

I also wanted to add namespaces, but I've got few time remaining...

I'll create issues with other ideas that I had/things that I got lazy to check.
